File: cli/src/main/kotlin/com/maestrorecorder/cli/merger/EventMerger.kt
package com.maestrorecorder.cli.merger

import com.maestrorecorder.cli.grpc.AgentClient
import com.maestrorecorder.shared.models.MaestroCommand
import com.maestrorecorder.shared.models.TapOnSelector
import com.maestrorecorder.shared.models.UiElement
import com.maestrorecorder.shared.models.UserAction
import kotlinx.coroutines.async
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.channelFlow
import kotlinx.coroutines.launch

class EventMerger(private val elementLookup: suspend (Int, Int) -> UiElement?) {

    constructor(agentClient: AgentClient) : this({ x, y ->
        try {
            agentClient.getElementAt(x, y)
        } catch (_: Exception) {
            null
        }
    })

    fun merge(actions: Flow<UserAction>): Flow<MaestroCommand> = channelFlow {
        coroutineScope {
            val pendingResults = Channel<MaestroCommand>(Channel.UNLIMITED)

            launch {
                actions.collect { action ->
                    when (action) {
                        is UserAction.Tap -> {
                            val command = async {
                                resolveTap(action)
                            }
                            pendingResults.send(command.await())

                            // Check if tap was on a text field — attempt text input detection
                            val textCommand = async { detectTextInput(action) }
                            val text = textCommand.await()
                            if (text != null) {
                                pendingResults.send(text)
                            }
                        }
                    }
                }
                pendingResults.close()
            }

            for (command in pendingResults) {
                send(command)
            }
        }
    }

    private suspend fun resolveTap(tap: UserAction.Tap): MaestroCommand {
        val element = elementLookup(tap.x, tap.y)

        val selector = if (element != null) {
            ElementSelector.selectBest(element)
        } else {
            TapOnSelector.ByPoint(tap.x, tap.y)
        }

        return MaestroCommand.TapOn(selector)
    }

    private suspend fun detectTextInput(tap: UserAction.Tap): MaestroCommand? {
        val element = elementLookup(tap.x, tap.y) ?: return null

        val isTextField = element.focused ||
            element.className?.contains("EditText") == true ||
            element.className?.contains("TextField") == true

        if (!isTextField) return null

        // Wait for user to type, then poll the field's text content
        delay(2000)

        val updatedElement = elementLookup(tap.x, tap.y) ?: return null
        val text = updatedElement.text

        if (!text.isNullOrBlank() && text != element.text) {
            return MaestroCommand.InputText(text)
        }

        return null
    }
}

File: packages/server/src/recording/event-merger.ts
import type { UserAction, MaestroCommand, UiElement } from '@maestro-recorder/shared';
import { AgentClient } from './agent-client.js';
import { selectBestSelector } from './element-selector.js';

export class EventMerger {
  constructor(
    private agent: AgentClient,
    private screenWidth: number,
    private screenHeight: number,
  ) {}

  async merge(action: UserAction): Promise<{ command: MaestroCommand; element: UiElement | null }> {
    switch (action.type) {
      case 'tap': return this.mergeTap(action);
      case 'longPress': return this.mergeLongPress(action);
      case 'swipe': return { command: this.mergeSwipe(action), element: null };
      case 'scroll': return { command: this.mergeScroll(action), element: null };
    }
  }

  private async mergeTap(action: { x: number; y: number }): Promise<{ command: MaestroCommand; element: UiElement | null }> {
    const element = await this.agent.elementAt(action.x, action.y);
    const selector = selectBestSelector(element, action.x, action.y);
    return { command: { type: 'tapOn', selector }, element };
  }

  private async mergeLongPress(action: { x: number; y: number }): Promise<{ command: MaestroCommand; element: UiElement | null }> {
    const element = await this.agent.elementAt(action.x, action.y);
    const selector = selectBestSelector(element, action.x, action.y);
    return { command: { type: 'longPressOn', selector }, element };
  }

  private mergeSwipe(action: { startX: number; startY: number; endX: number; endY: number }): MaestroCommand {
    const start = this.toPercent(action.startX, action.startY);
    const end = this.toPercent(action.endX, action.endY);
    return { type: 'swipe', start, end } as MaestroCommand;
  }

  private mergeScroll(action: { direction: string }): MaestroCommand {
    return { type: 'scroll' };
  }

  /** Detect if a tap landed on a text field and capture subsequent text input */
  async detectTextInput(element: UiElement | null, x: number, y: number): Promise<string | null> {
    if (!element) return null;

    const isTextField =
      element.focused ||
      element.className?.includes('EditText') === true ||
      element.className?.includes('TextField') === true;

    if (!isTextField) return null;

    const originalText = element.text || '';

    // Wait for user to type, then poll
    await sleep(2000);

    const updated = await this.agent.elementAt(x, y);
    if (!updated?.text || updated.text === originalText) return null;

    return updated.text;
  }

  private toPercent(x: number, y: number): string {
    const px = Math.round((x / this.screenWidth) * 100);
    const py = Math.round((y / this.screenHeight) * 100);
    return `${px}%, ${py}%`;
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise(r => setTimeout(r, ms));
}
